lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1417 @@
+    $qemuimgPath = Join-Path $BIN_DIR "qemu-img.exe"
+    if (!(Test-Path $mkisofsPath) -or !(Test-Path $qemuimgPath)) {
+        Write-JujuLog "Extracting OpenStack binaries"
+        $zipPath = Join-Path $FILES_DIR "openstack_bin.zip"
+        Expand-ZipArchive $zipPath $BIN_DIR
+    }
+  
+    $networkType = Get-JujuCharmConfig -Scope 'network-type'
+    Initialize-GitRepositories $networkType $BranchName $BuildFor
+
+    $standardProjects = 'openstack/nova', 'openstack/neutron', 'openstack/os-win'
+    if ($buildFor -in $standardProjects) {
+        $defaultProjects = $standardProjects -notmatch $buildFor
+        foreach ($project in $defaultProjects) {
+            Install-Projects $project
+        }
+        Install-NetworkType
+        Install-Projects $buildFor
+    }
+    else {
+        foreach ($project in $standardProjects) {
+            Install-Projects $project
+        }
+        Install-NetworkType
+        if (($buildFor -ne "none") -and ($buildFor -ne "openstack/networking-hyperv")) {
+            Install-Projects $buildFor
+        }
+    }
+
+    Write-JujuLog "Environment initialization done."
+}
+
+
+function New-OpenStackService {
+    Param(
+        [string]$ServiceName,
+        [string]$ServiceDescription,
+        [string]$ServiceExecutable,
+        [string]$ServiceConfig,
+        [string]$ServiceUser,
+        [string]$ServicePassword
+    )
+
+    $filter = "Name='{0}'" -f $ServiceName
+
+    $service = Get-WmiObject -Namespace "root\cimv2" -Class Win32_Service -Filter $filter
+    if($service) {
+        Write-JujuLog "Service $ServiceName is already created."
+        return $true
+    }
+
+    $serviceFileName = "OpenStackService.exe"
+    if(!(Test-Path "$SERVICE_DIR\$serviceFileName")) {
+        Copy-Item "$FILES_DIR\$serviceFileName" "$SERVICE_DIR\$serviceFileName"
+    }
+
+    New-Service -Name "$ServiceName" -DisplayName "$ServiceName" -Description "$ServiceDescription" -StartupType "Manual" `
+                -BinaryPathName "$SERVICE_DIR\$serviceFileName $ServiceName $ServiceExecutable --config-file $ServiceConfig" `
+
+    if((Get-Service -Name $ServiceName).Status -eq "Running") {
+        Stop-Service $ServiceName
+    }
+
+    Set-ServiceLogon -Services @($ServiceName) -UserName $ServiceUser -Password $ServicePassword
+}
+
+
+function Watch-ServiceStatus {
+    Param(
+        [string]$ServiceName,
+        [int]$IntervalSeconds
+    )
+
+    $count = 0
+    while ($count -lt $IntervalSeconds) {
+        if ((Get-Service -Name $ServiceName).Status -ne "Running") {
+            $count += 1
+            Start-Sleep -Seconds 1
+            continue
+        }
+        return
+    }
+    Throw "$ServiceName has errors. Please check the logs."
+}
+
+
+function Get-JujuVMSwitchName {
+    $VMswitchName = Get-JujuCharmConfig -Scope "vmswitch-name"
+    if (!$VMswitchName){
+        return "br100"
+    }
+    return $VMswitchName
+}
+
+
+function Get-InterfaceFromConfig {
```

Comment:
So, I see that there are some common functions with the upstream `nova-hyperv` charm.

Some commits ago, I've created 2 separated modules for [OVS](https://github.com/cloudbase/nova-hyperv-charm/tree/master/lib/Modules/OVSCharmUtils) (this is made to work with the stable MSI though and I'm not sure if you can use it straight away) and [HyperVNetworking](https://github.com/cloudbase/nova-hyperv-charm/tree/master/lib/Modules/HyperVNetworking).

This `Get-InterfaceFromConfig` function you can import from the `HyperVNetworking` module. I would suggest we maintain common modules in a separate repo (I've started one mentioned above) and we just sync them in the charms that uses them. This way, we have no duplicate code and we maintain code in one place.